### PR TITLE
fix(cli): discard changes if lb4 `relation` fails. Allow customized relation name for belongsTo

### PR DESCRIPTION
**Pull request by agnes512**
Fixes https://github.com/strongloop/loopback-next/issues/3287 and https://github.com/strongloop/loopback-next/issues/4209

1st commit: code changes
2nd commit: tests

#### Example: 

- **hasMany**
? SourceModel: `Customer`
? DestModel: `Order`,
? Foreign key to define on the target model: `customerId`
? Source property name for the relation getter: `orders
? Allow Order queries to include data from related Customer instances? Yes
==> issues:
  - if generation gets aborted, part of code still gets generated in Model and Repository 
  - the default value doesn't match with the code in `/repository/relation`

- **belongTo**
? SourceModel: `Order`
? DestModel: `Customer`,
? Source property name for the relation getter: `customerId`
? Allow ..inclusion? Yes
==> issues:
  - if generation gets aborted, part of code still gets generated in Model and Repository
  - the default value doesn't match with the code in `/repository/relation` 
  - generated code failed if the foreign key name is not the default value
  - generator doesn't respect custom relation name, the only relation it generates is `targetModelName` (e.g `customer`)

## New changes

- checks if corresponding repository exists
- checks if pk exists and validates the name
- checks if fk exists and validates the name
- asks fk for both hasMany and belongsTo
- asks relation name for both hasMany and belongsTo
- checks if the relation name exists
- allow custom relation name for belongsTo
  - This fixes https://github.com/strongloop/loopback-next/issues/4209
  - customized relation name will be specified in the relation metadata

#### Examples of new prompt:

- **hasMany**
? Please select the relation type `hasMany`
? Please select source model `Customer`
? Please select target model `Order`
? Foreign key name to define on the _target_ model `c_id` // not default value
? Source property name for the relation getter (will be the relation name) `orders`
? Allow Customer queries to include data from related Review instances? Yes

=>
```
// Customer property:
  @hasMany(() => Order, {keyTo: 'c_id'})  // custom fk
  orders: Order[];

// Order fk:
  @property({
    type: 'number',
  })
  c_id?: number;  // custom fk
```

- **belongsTo**
? Please select the relation type `belongsTo`
? Please select source model `Order`
? Please select target model `Customer`
? Foreign key name to define on the _source_ model `c_id`
? Relation name my_customer   // not default name
? Allow Customer queries to include data from related Review instances? Yes

=> 
```
//Order fk:
  @belongsTo(() => Customer, {name: 'my_customer'}) . // allow custom relation name
  c_id: number;    // allow custom fk name
```

For belongsTo, if the fk name is the same as the relation name, the prompt would ask the user to re-enter a new name to avoid a known issue: [Relations with custom names not working ](https://github.com/strongloop/loopback-next/issues/4392)

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/cli/generators/relation/belongs-to-relation.generator.js [hunk 1]
```diff
@@ -101,10 +111,36 @@ module.exports = class BelongsToRelationGenerator extends BaseRelationGenerator
...
+  ) {
+    // checks if relation name is customized
+    let relationDecorator = [
+      {
+        name: 'belongsTo',
+        arguments: [`() =>  ${className}`],
+      },
+    ];
+    // except the customized relation name, also need to check if the property name is the same as relation name
+    // or it will get 'navigational property' error when creating instances
```
Comment (agnes512) [inline packages/cli/generators/relation/belongs-to-relation.generator.js]: related [stackoverflow question](https://stackoverflow.com/questions/59468329/unhandled-error-in-post-order-products-500-error-navigational-properties-are).

File: packages/cli/generators/relation/index.js [hunk 2]
```diff
@@ -267,24 +314,69 @@ module.exports = class RelationGenerator extends ArtifactGenerator {
...
+      // should check the existance of fk here.
+      // default source key for belongsTo
+      this.artifactInfo.defaultSourceKeyName =
+        utils.camelCase(this.artifactInfo.destinationModel) +
+        utils.toClassName(this.artifactInfo.destinationModelPrimaryKey);
+
+      return this.prompt([
+        {
+          type: 'string',
+          name: 'sourceKeyName',
```
Comment (agnes512) [inline packages/cli/generators/relation/index.js]: use `sourceKeyName`, `relationName`, and `defaultRelationName` to check if the source key of belongsTo is customized. These three properties also check if the `name` field of the relation metadata needs to be specified.

File: packages/cli/generators/relation/index.js [hunk 3]
```diff
@@ -338,50 +430,64 @@ module.exports = class RelationGenerator extends ArtifactGenerator {
...
     if (this.options.relationName) {
       debug(
         `Relation name received from command line: ${this.options.relationName}`,
       );
       this.artifactInfo.relationName = this.options.relationName;
     }
+    this.artifactInfo.defaultRelationName = this._getDefaultRelationName();
+    // for hasMany && hasOne, the source key is the same as the relation name
+    const msg =
+      this.artifactInfo.relationType === 'belongsTo'
```
Comment (agnes512) [inline packages/cli/generators/relation/index.js]: This property name will be the source key and the relation name for hasMany and hasOne. So I separate it from belongsTo here.

**Comment by agnes512** [comment]
@derdeka @achrinza could you take a look? part of code fixes https://github.com/strongloop/loopback-next/issues/4209. Thanks.

File: packages/cli/generators/relation/base-relation.generator.js [hunk 4]
```diff
@@ -98,23 +98,17 @@ module.exports = class BaseRelationGenerator extends ArtifactGenerator {
...
     };
 
-    if (relationUtils.doesPropertyExist(classDeclaration, property.name)) {
-      throw new Error(
-        'property ' + property.name + ' already exist in the repository.',
-      );
-    } else {
-      relationUtils.addProperty(classDeclaration, property);
-    }
+    // already chcked the existence of property before
```
Comment (bajtos) [inline packages/cli/generators/relation/base-relation.generator.js]: Typo.

```suggestion
    // already checked the existence of property before
```

File: packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js [hunk 5]
```diff
@@ -255,7 +259,7 @@ export class Order extends Entity {
   name?: string;
 
   @belongsTo(() => Customer)
-  myRelation: number;
+  customerId: number;
```
Comment (bajtos) [inline packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js]: Is this change intentional and legitimate? I have no idea, just checking to ensure we don't accidentally overlook a regression.

Comment (agnes512) [reply packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js]: This test aims to check if the pre-exist property `@property{} something;` will be overwritten by `@belongsTo` decorator. That's why I use the default relation name to make the intention more clear.

File: packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js [hunk 6]
```diff
@@ -46,12 +50,12 @@ export class OrderClassRepository extends DefaultCrudRepository<
   typeof OrderClass.prototype.orderNumber
 > {
 
-  public readonly customerClass: BelongsToAccessor<CustomerClass, typeof OrderClass.prototype.orderNumber>;
+  public readonly customerClassCustNumber: BelongsToAccessor<CustomerClass, typeof OrderClass.prototype.orderNumber>;
```
Comment (jannyHou) [inline packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js]: hmm what does `Cust` mean 🤔 ?

File: packages/cli/test/integration/generators/relation.has-many.integration.js [hunk 7]
```diff
@@ -96,6 +98,38 @@ describe('lb4 relation HasMany', function() {
     );
   });
 
+  context('Execute relation with exist relation name', () => {
```
Comment (jannyHou) [inline packages/cli/test/integration/generators/relation.has-many.integration.js]: nitpick: `existing`

File: packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js (hunk 6 above)
Comment (agnes512) [reply packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js]: `CustNumber` is the name of the id property of `customerClass` model.

File: packages/cli/generators/relation/index.js [hunk 8]
```diff
@@ -338,50 +430,64 @@ module.exports = class RelationGenerator extends ArtifactGenerator {
...
+      this.artifactInfo.srcRepositoryFileObj = new relationUtils.AstLoopBackProject().addExistingSourceFile(
+        this.artifactInfo.srcRepositoryFile,
+      );
+
+      const repoClassDeclaration = this.artifactInfo.srcRepositoryFileObj.getClassOrThrow(
+        this.artifactInfo.srcRepositoryClassName,
+      );
+      // checks if the relation name already exists in repo
+      if (
+        relationUtils.doesPropertyExist(
```
Comment (jannyHou) [inline packages/cli/generators/relation/index.js]: Just curious: I think the relation definition is in the model file, why do we have to use more steps to get the corresponding repository file and check the existence?

File: packages/cli/generators/relation/index.js [hunk 9]
```diff
@@ -267,24 +314,69 @@ module.exports = class RelationGenerator extends ArtifactGenerator {
...
+      this.artifactInfo.defaultSourceKeyName =
+        utils.camelCase(this.artifactInfo.destinationModel) +
+        utils.toClassName(this.artifactInfo.destinationModelPrimaryKey);
+
+      return this.prompt([
+        {
+          type: 'string',
+          name: 'sourceKeyName',
+          message: PROMPT_MESSAGE_SOURCE_KEY_NAME,
+          when: this.artifactInfo.relationName === undefined,
```
Comment (jannyHou) [inline packages/cli/generators/relation/index.js]: nitpick: `when: !this.artifactInfo.relationName`

File: packages/cli/test/integration/generators/relation.has-many.integration.js (hunk 7 above)
Comment (emonddr) [inline packages/cli/test/integration/generators/relation.has-many.integration.js]: with `existing` relation

File: packages/cli/test/integration/generators/relation.has-many.integration.js [hunk 10]
```diff
@@ -96,6 +98,38 @@ describe('lb4 relation HasMany', function() {
...
+                SourceEntries.CustomerModel,
+                SourceEntries.OrderModel,
+                SourceEntries.CustomerRepository,
+                SourceEntries.OrderRepository,
+              ],
+            }),
+          )
+          .withPrompts(prompt),
+      ).to.be.rejectedWith(
+        `relation myCustomer already exist in the repository OrderRepository.`,
```
Comment (emonddr) [inline packages/cli/test/integration/generators/relation.has-many.integration.js]: already `exists`

**Comment by emonddr** [comment]
The relation-based test cases pass, so that's great.
Some coverage fell though :S

File: packages/cli/generators/relation/index.js (hunk 8 above)
Comment (agnes512) [reply packages/cli/generators/relation/index.js]: Good question! From my understanding the model has _the property_ (e.g `orders` or `customerId`). And the repository has _the relation name_ (e.g `orders` or `customer`). They might have the same name but most of time they don't. For example, for belongsTo relation, the relation name is `customer`.

The generator throws errors if the repository already has the relation name defined. And the generator didn't have such checks before .

File: packages/cli/generators/relation/belongs-to-relation.generator.js [hunk 11]
```diff
@@ -140,22 +176,22 @@ module.exports = class BelongsToRelationGenerator extends BaseRelationGenerator
...
 
   _addCreatorToRepositoryConstructor(classConstructor) {
-    const relationPropertyName = this._getRepositoryRelationPropertyName();
+    const relationName = this.artifactInfo.relationName;
     const statement =
-      `this.${relationPropertyName} = ` +
+      `this.${relationName} = ` +
       `this.createBelongsToAccessorFor('` +
-      `${this.artifactInfo.relationName.replace(/Id$/, '')}',` +
+      `${relationName}',` +
```
[the remark this answers is not in the document]
Comment (agnes512) [reply packages/cli/generators/relation/belongs-to-relation.generator.js]: I have `relationName` defined above so we don't need it here anymore.

File: packages/cli/generators/relation/has-many-relation.generator.js [hunk 12]
```diff
@@ -166,7 +168,7 @@ module.exports = class HasManyRelationGenerator extends BaseRelationGenerator {
   }
 
   _getRepositoryRelationPropertyName() {
-    return utils.pluralize(utils.camelCase(this.artifactInfo.dstModelClass));
+    return this.artifactInfo.relationName;
```
[the remark this answers is not in the document]
Comment (agnes512) [reply packages/cli/generators/relation/has-many-relation.generator.js]: I have relationName defined above so we can just use the value here.

File: packages/cli/generators/relation/index.js [hunk 13]
```diff
@@ -134,6 +156,16 @@ module.exports = class RelationGenerator extends ArtifactGenerator {
...
     }
+    let repoList;
+    try {
+      debug(`repository list dir ${this.artifactInfo.repoDir}`);
+      repoList = await utils.getArtifactList(
+        this.artifactInfo.repoDir,
+        'repository',
+      );
+    } catch (err) {
+      return this.exit(err);
```
[the remark this answers is not in the document]
Comment (agnes512) [reply packages/cli/generators/relation/index.js]: We use exit for generators

**Comment by agnes512** [comment]
> The relation-based test cases pass (i ran them on my laptop), so that's great. Doesn't appear to be any regressions.
> Some coverage fell though :S

If you check the tests It is caused by the generator every time it does `exit(err)`, not the actual code itself. I don't any way to skip it. If anyone knows please let me know :smile:

**Comment by agnes512** [comment]
> Just curious; do we also need to update the docs that reference the CLI prompt procedure?
> 
> See:
> 
> * https://loopback.io/doc/en/lb4/Relation-generator.html
> * https://loopback.io/doc/en/lb4/todo-list-tutorial-relations.html

Yes, I will add them once the proposal looks good to most of you :smile:

File: packages/cli/generators/relation/index.js (hunk 8 above)
Comment (jannyHou) [reply packages/cli/generators/relation/index.js]: >  And the repository has the relation name

Oh that's smart! 👍 makes perfect sense then

File: packages/cli/generators/relation/base-relation.generator.js (hunk 4 above)
Comment (jannyHou) [inline packages/cli/generators/relation/base-relation.generator.js]: (sorry I am not very familiar with the relation cli which results in many questions ❔ ❔ ) Hmm, where is it checked? did a quick search of code before it but didn't find any clue 🙈

File: packages/cli/generators/relation/base-relation.generator.js [hunk 14]
```diff
@@ -98,23 +98,17 @@ module.exports = class BaseRelationGenerator extends ArtifactGenerator {
...
 
   _addParametersToRepositoryConstructor(classConstructor) {
     const parameterName =
       utils.camelCase(this.artifactInfo.dstRepositoryClassName) + 'Getter';
 
     if (relationUtils.doesParameterExist(classConstructor, parameterName)) {
-      throw new Error(
-        'Parameter ' + parameterName + ' already exist in the constructor.',
-      );
+      // no need to check if the getter already exists
```
Comment (jannyHou) [inline packages/cli/generators/relation/base-relation.generator.js]: Doesn't parameter exist imply the relation already exists? And therefore an error should be thrown like the deleted code?

File: packages/cli/generators/relation/belongs-to-relation.generator.js [hunk 15]
```diff
@@ -101,10 +111,32 @@ module.exports = class BelongsToRelationGenerator extends BaseRelationGenerator
...
+        name: 'belongsTo',
+        arguments: [`() =>  ${className}`],
+      },
+    ];
+    // already checked if the relation name is the same as the source key before
+    if (defaultRelationName !== relationName) {
+      relationDecorator = [
+        {
+          name: 'belongsTo',
+          arguments: [`() =>  ${className}, {name: '${relationName}'}`],
```
Comment (jannyHou) [inline packages/cli/generators/relation/belongs-to-relation.generator.js]: I am a bit confused about `defaultRelationName` and `sourceKeyName`, wouldn't they be the same? 
I find it hard to think of a user scenario that `defaultRelationName`, `sourceKeyName` and `relationName` are all different...

File: packages/cli/generators/relation/belongs-to-relation.generator.js [hunk 16]
```diff
@@ -122,7 +154,7 @@ module.exports = class BelongsToRelationGenerator extends BaseRelationGenerator
   }
 
   _getRepositoryRelationPropertyName() {
-    return utils.camelCase(this.artifactInfo.dstModelClass);
+    return this.artifactInfo.relationName;
```
Comment (jannyHou) [inline packages/cli/generators/relation/belongs-to-relation.generator.js]: I guess this line fixes https://github.com/strongloop/loopback-next/issues/4209?

File: packages/cli/generators/relation/index.js [hunk 17]
```diff
@@ -23,14 +23,19 @@ const ERROR_SOURCE_MODEL_PRIMARY_KEY_DOES_NOT_EXIST =
   'Source model primary key does not exist.';
 const ERROR_DESTINATION_MODEL_PRIMARY_KEY_DOES_NOT_EXIST =
   'Target model primary key does not exist.';
+const ERROR_REPOSITORY_DOES_NOT_EXIST =
+  'class does not exist. Please create repositories first with "lb4 repository" command.';
```
Comment (jannyHou) [inline packages/cli/generators/relation/index.js]: nitpick: `create **repository**`, instead of `create **repositories**`

File: packages/cli/generators/relation/base-relation.generator.js (hunk 4 above)
Comment (agnes512) [reply packages/cli/generators/relation/base-relation.generator.js]: Here the property is the relation name, which has been checked in [`index file`](https://github.com/strongloop/loopback-next/pull/4373/files#diff-d9d0cb4d78404ced50ce6c783a09ccbaR488). Sorry the PR is big, I should've separated them.

File: packages/cli/generators/relation/base-relation.generator.js (hunk 14 above)
Comment (agnes512) [reply packages/cli/generators/relation/base-relation.generator.js]: The param is getter, such as `@repository.getter('OrderRepository') protected orderRepositoryGetter: Getter<OrderRepository>`. And I don't think this implies a certain _relation name_ already exists. That's why I removed it. 

And I was also thinking about some edge cases. For example, since we allow customized `keyFrom`, user could have two _same_ relations with 2 models with two different names. In this case, they will share the same getter.

File: packages/cli/generators/relation/belongs-to-relation.generator.js (hunk 15 above)
Comment (agnes512) [reply packages/cli/generators/relation/belongs-to-relation.generator.js]: Good point. I will list out the differences:

Before, the generator asked `foreign key`, `relationName` for **hasMany**.  
 and it only asked `relationName` for **belongTo** if `defaultRelationName` doesn't exist. It used `relationName` as the foreign key name (e.g `customerId`). Then it use the target model as the _real relation name_ (e.g `customer).

This is why how it dealt with **belongsTo** is problematic:
1. if the default foreign key already exists, it didn't prompt
2. if the default foreign key doesn't exists, it takes whatever it is, but the _real relation name_ is still the target model, and this is different from how the `belongsToAccessor` generates the default name. For example: if you type in `customer_id` as your foreign key, the generator would have `customer` as the default relation name. But for the `belongsToAccessor`, the default relation name is `customer_id` instead of `customer`. That's why users are getting `Cannot find type of undefined` errors. 

In order to allow custom fk and relation name for belongsTo, I added a new field `sourceKeyName` to represent the foreign key of belongsTo. And `relationName` to represent the _real relation name_. 

The line `defaultRelationName !== relationName` here is to check if the relationName is customized. If so, it should be specified in the `name` field, just like how we documented in our docs [Relation Metadata](https://loopback.io/doc/en/lb4/BelongsTo-relation.html#relation-metadata).

Therefore, it can have correct property:
```
  @belongsTo(() => Customer, {name: 'my_customer'}) // customized relation name
  customer_id: number; // customized fk
```

File: packages/cli/generators/relation/belongs-to-relation.generator.js (hunk 16 above)
Comment (agnes512) [reply packages/cli/generators/relation/belongs-to-relation.generator.js]: See comment https://github.com/strongloop/loopback-next/pull/4373/files#r365448472. This line is part of fix.

File: packages/cli/lib/utils.js [hunk 18]
```diff
@@ -140,6 +140,84 @@ exports.validateNotExisting = function(projDir) {
   return true;
 };
 
+/**
+ * validate source key or foreign key for relations
+ */
```
Comment (nabdelgadir) [inline packages/cli/lib/utils.js]: Call `/* istanbul ignore next */` here instead?

File: packages/cli/test/fixtures/relation/index.js [hunk 19]
```diff
@@ -83,12 +83,24 @@ const SourceEntries = {
     content: readSourceFile('./repositories/order-class-type.repository.ts'),
   },
 
-  NoKeyModel: {
+  NokeyModel: {
```
Comment (nabdelgadir) [inline packages/cli/test/fixtures/relation/index.js]: I think this should stay as `NoKeyModel`

File: packages/cli/generators/relation/belongs-to-relation.generator.js [hunk 20]
```diff
@@ -140,22 +172,22 @@ module.exports = class BelongsToRelationGenerator extends BaseRelationGenerator
   }
 
   _addCreatorToRepositoryConstructor(classConstructor) {
-    const relationPropertyName = this._getRepositoryRelationPropertyName();
+    const relationName = this.artifactInfo.relationName;
```
[the remark this answers is not in the document]
Comment (agnes512) [reply packages/cli/generators/relation/belongs-to-relation.generator.js]: Yes, just personal preference as I have it defined above.

File: packages/cli/test/fixtures/relation/index.js (hunk 19 above)
Comment (agnes512) [reply packages/cli/test/fixtures/relation/index.js]: I want to make it consistent as the model name is `Nokey` in `nokey.model.ts`.

File: packages/cli/generators/relation/base-relation.generator.js (hunk 14 above)
Comment (jannyHou) [reply packages/cli/generators/relation/base-relation.generator.js]: > since we allow customized keyFrom, user could have two same relations with 2 models with two different names. In this case, they will share the same getter.

👍 Thank you I agree with you.
The property should be checked instead of parameter.

File: packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js (hunk 6 above)
Comment (agnes512) [reply packages/cli/snapshots/integration/generators/relation.belongs-to.integration.snapshots.js]: I change the generator to generate `customerClassId` instead to match with our code in `/repository/relation`

File: packages/cli/lib/utils.js (hunk 18 above)
Comment (agnes512) [reply packages/cli/lib/utils.js]: Right, it's so much better, thanks!

File: packages/cli/test/fixtures/relation/index.js (hunk 19 above)
Comment (nabdelgadir) [reply packages/cli/test/fixtures/relation/index.js]: Should we update the one in `nokey.model.ts` then?

File: docs/site/BelongsTo-relation.md [hunk 21]
```diff
@@ -43,6 +43,11 @@ related routes, you need to perform the following steps:
 This section describes how to define a `belongsTo` relation at the model level
 using the `@belongsTo` decorator to define the constraining foreign key.
 
+LB4 also provides an CLI tool `lb4 relation` to generate `belongsTo` relation
```
Comment (nabdelgadir) [inline docs/site/BelongsTo-relation.md]: So that the reader doesn't have to come back to this section, can you move it to after the `Relation Metadata` section?

File: docs/site/Relation-generator.md [hunk 22]
```diff
@@ -32,7 +32,8 @@ lb4 relation [options]
 - `--relationType`: Relation type.
 - `--sourceModel`: Source model.
 - `--destinationModel`: Destination model.
-- `--foreignKeyName`: Destination model foreign key name.
+- `--foreignKeyName`: Destination (Source) model foreign key name for HasMany
+  (BelongsTo) relation.
```
Comment (nabdelgadir) [inline docs/site/Relation-generator.md]: suggestion: `Destination/Source model foreign key name for HasMany/BelongsTo relation, respectively.` (feel free to ignore)

File: packages/cli/test/fixtures/relation/index.js (hunk 19 above)
Comment (agnes512) [reply packages/cli/test/fixtures/relation/index.js]: either is fine..

File: docs/site/BelongsTo-relation.md (hunk 21 above)
Comment (agnes512) [reply docs/site/BelongsTo-relation.md]: I was thinking where is the best place to put it. I don't think after relation metadata is a good idea because we introduce how users can customize keyFrom, keyTo, relation name, and even column name, and most of them are not supported by the CLI. That's why I put it here, right after we introduce the basic idea of each relation.
